test/gtest/uct/ib/test_sockaddr.cc

```diff
@@ -434,35 +469,106 @@ class test_uct_cm_sockaddr : public uct_test {
+        cm_start_listen();
+        m_client->connect(0, *m_server, 0, m_connect_addr, client_cm_priv_data_cb,
+                          client_connect_cb, client_disconnect_cb, this);
+
+        wait_for_bits(&m_cm_state, TEST_CM_STATE_CONNECT_REQUESTED);
+        EXPECT_TRUE(m_cm_state & TEST_CM_STATE_CONNECT_REQUESTED);
+    }
+
+    size_t priv_data_do_pack(size_t pack_limit, void *priv_data)
+    {
+        if (pack_limit < long_priv_data_len) {
+            /* a transport with a small private data length (rdmacm_cm) */
+            memcpy(priv_data, short_priv_data.data(), short_priv_data_len);
+            return short_priv_data_len;
+        } else {
+            /* a transport with a large private data length (tcp_sockcm) */
+            memcpy(priv_data, long_priv_data.data(), long_priv_data_len);
+            return long_priv_data_len;
+        }
+    }
+
+    ssize_t common_priv_data_cb(void *arg, size_t pack_limit, void *priv_data)
     {
         test_uct_cm_sockaddr *self = reinterpret_cast<test_uct_cm_sockaddr *>(arg);
         size_t priv_data_len;
 
-        priv_data_len = uct_test::entity::priv_data_do_pack(priv_data);
-        EXPECT_LE(priv_data_len, self->m_client->max_conn_priv);
+        priv_data_len = self->priv_data_do_pack(pack_limit, priv_data);
+        EXPECT_LE(priv_data_len, pack_limit);
         return priv_data_len;
     }
 
-    void cm_listen_and_connect() {
-        skip_tcp_sockcm();
-        cm_start_listen();
-        m_client->connect(0, *m_server, 0, m_connect_addr, client_cm_priv_data_cb,
-                          client_connect_cb, client_disconnect_cb, this);
+    static ssize_t client_cm_priv_data_cb(void *arg,
+                                          const uct_cm_ep_priv_data_pack_args_t
+                                          *pack_args, void *priv_data)
+    {
+        test_uct_cm_sockaddr *self = reinterpret_cast<test_uct_cm_sockaddr *>(arg);
+        return self->common_priv_data_cb(arg, self->m_client->max_conn_priv, priv_data);
+    }
 
-        wait_for_bits(&m_cm_state, TEST_CM_STATE_CONNECT_REQUESTED);
-        EXPECT_TRUE(m_cm_state & TEST_CM_STATE_CONNECT_REQUESTED);
+    static ssize_t server_cm_priv_data_cb(void *arg,
+                                          const uct_cm_ep_priv_data_pack_args_t
+                                          *pack_args, void *priv_data)
+    {
+        test_uct_cm_sockaddr *self = reinterpret_cast<test_uct_cm_sockaddr *>(arg);
+        return self->common_priv_data_cb(arg, self->m_server->max_conn_priv, priv_data);
+    }
+
+    void accept(uct_cm_h cm, uct_conn_request_h conn_request,
+                uct_cm_ep_server_conn_notify_callback_t notify_cb,
+                uct_ep_disconnect_cb_t disconnect_cb,
+                void *user_data)
+    {
+        uct_ep_params_t ep_params;
+        ucs_status_t status;
+        uct_ep_h ep;
+
+        ASSERT_TRUE(m_server->listener());
+        m_server->reserve_ep(m_server->num_eps());
```

Comment:
why we can't just push back when the ep is created?

Reply:
in order to use reset() to set the dtor